chore(docs): documenting reference, how-to and explainer on the usage of oracles (#4002)
noir-lang/noir: docs/docs/how_to/how-to-oracles.md

---
title: How to use Oracles
description: Learn how to use oracles in your Noir program with examples in both Nargo and NoirJS. This guide also covers writing a JSON RPC server and providing custom foreign call handlers for NoirJS.
keywords:
  - Noir Programming
  - Oracles
  - Nargo
  - NoirJS
  - JSON RPC Server
  - Foreign Call Handlers
sidebar_position: 1
---

> This guide shows you how to use oracles in your Noir program. For the sake of clarity, it is assumed that:

Reviewer: ```suggestion
This guide shows you how to use oracles in your Noir program. For the sake of clarity, it is assumes that:
```

> - You have read the [explainer on Oracles](../explainers/explainer-oracle.md) and are comfortable with the concept

Reviewer: for consistency
```suggestion
- You have read the [explainer on Oracles](../explainers/explainer-oracle.md) and are comfortable with the concept.
```

> - You already have your own NoirJS and/or Nargo app. If you don't, you can use the [vite-hardhat starter](https://github.com/noir-lang/noir-starter/tree/main/vite-hardhat) as a boilerplate (for example, by using its [devcontainer in codespaces](./using-devcontainers.mdx))

Reviewer: ```suggestion
- You already have your own NoirJS and/or Nargo app. If you don't, you can use the [vite-hardhat starter](https://github.com/noir-lang/noir-starter/tree/main/vite-hardhat) as a boilerplate (for example, by using its [devcontainer in codespaces](./using-devcontainers.mdx)).
```

- You understand the concept of a JSON-RPC server. Visit the [JSON-RPC website](https://www.jsonrpc.org/) if you need a refresher.
- You are comfortable with server-side javascript. Will skip any details on installing Node, packages, etc, so as to keep the guide short and straight to the point.

> For reference, you can find the snippets used in this tutorial on the [Aztec DevRel Repository](https://github.com/AztecProtocol/dev-rel/tree/main/how_to_oracles/code-snippets/how-to-oracles)

Reviewer: ```suggestion
For reference, you can find the snippets used in this tutorial on the [Aztec DevRel Repository](https://github.com/AztecProtocol/dev-rel/tree/main/how_to_oracles/code-snippets/how-to-oracles).
```

## Rundown

This guide has 3 major steps:

> 1. How to modify our circuit to make use of oracle calls as unconstrained functions.

Reviewer: ```suggestion
1. How to modify our circuit to make use of oracle calls as unconstrained functions
```

2. How to write a JSON RPC Server to resolve these oracle calls with Nargo
3. How to use them in Nargo and how to provide a custom resolver in NoirJS

## Step 1 - Modify your circuit

An oracle is defined in a circuit by defining two methods:

> - An unconstrained method - This tells the Noir compiler that it is executing a Brillig block.

Reviewer: Should you add a link to "Brillig" for those that are not familiar with the term to learn more?

Reviewer 2: ```suggestion
- An unconstrained method - This tells the compiler that it is executing an [unconstrained functions](../noir/concepts//unconstrained.md).
```

Author: Perfect @Savio-Sou but I still see so much "brillig" everywhere, like debugger, etc. It's a discussion for another time

- A decorated oracle method - This tells the compiler that this method is an RPC call.

An example of an oracle that returns a `Field` would be:

```rust
#[oracle(getSquared)]
unconstrained fn squared(number: Field) -> Field { }

unconstrained fn get_squared(number: Field) -> Field {
    squared(number)
}
```

In this example, we're wrapping our oracle function in a unconstrained method, and decorating it with `oracle(getSquared)`. We can then call the unconstrained function as we would call any other function:

```rust
fn main(input: Field) {

>     let squared = get_squared(input);

Reviewer: Is it required to have the `squared` function wrapped in the `get_squared` function? We can't just call `squared` directly in the `main` fn? if so, why?

Author: Just followed the tests here, apparently you need to wrap the oracle in an unconstrained function. I get a warning message telling me that, too.

Not amazing but I would guess the decorator runs a macro or something that transforms the function into an oracle.

}
```

> In the next section, we will make this `getSquared` be a method of the RPC server Noir will use.

Reviewer: It'd be useful to call out that the RPC server will use the method defined in the decorator, not the name of the Noir function. This confused me for more than a minute.

```suggestion
In the next section, we will make this `getSquared` (defined on the `squared` decorator) be a method of the RPC server Noir will use.

```

If you're using Typescript, the following types may be helpful in understanding the expected return value and making sure they're easy to follow:

```js
interface Value {
  inner: string,
}

interface SingleForeignCallParam {
  Single: Value,
}

interface ArrayForeignCallParam {
  Array: Value[],
}

type ForeignCallParam = SingleForeignCallParam | ArrayForeignCallParam;

interface ForeignCallResult {
  values: ForeignCallParam[],
}
```

:::

## Step 3 - Usage with Nargo

Using the [`nargo` CLI tool](../getting_started/installation/index.md), you can use oracles in the `nargo test`, `nargo execute` and `nargo prove` commands by passing a value to `--oracle-resolver`. For example:

```bash
nargo test --oracle-resolver http://localhost:5555
```

This tells `nargo` to use your RPC Server URL whenever it finds an oracle decorator.

## Step 4 - Usage with NoirJS

In a JS environment, an RPC server is not strictly necessary, as you may want to resolve your oracles without needing any JSON call at all. NoirJS simply expects that you pass a callback function when you generate proofs, and that callback function can be anything.

> For example, if your circuit expects the host machine to provide CPU pseudo-randomness, you could simply pass it as the `foreignCallHandler`. You don't strictly need to create an RPC server to serve pseudo-randomness, as you may as well get it directly in your app:

Reviewer 2: ```suggestion
For example, if your Noir program expects the host machine to provide CPU pseudo-randomness, you could simply pass it as the `foreignCallHandler`. You don't strictly need to create an RPC server to serve pseudo-randomness, as you may as well get it directly in your app:
```

```js
const foreignCallHandler = (name, inputs) => crypto.randomBytes(16) // etc

await noir.generateFinalProof(inputs, foreignCallHandler)
```

> As one can see, in NoirJS, the [`foreignCallHandler`](../reference/NoirJS/noir_js/type-aliases/ForeignCallHandler.md) function simply means "a callback function that returns a value of type [`ForeignCallOutput`](../reference/NoirJS/noir_js/type-aliases/ForeignCallOutput.md). It doesn't have to be an RPC call like in the case for Nargo.

Reviewer: These two links can't be resolved when i run the docs locally.

Reviewer: Do i need to run a command or something locally so they are generated before starting the docusaurus server? If so, it'd be useful to add the command to `yarn start` in the docs package.json

Author: Yes these point to autogenerated docs but they should be generated with yarn start... I'll investigate a bit

Author: Fixed. Issue happened because I was rebuilding _just_ the docs, thus it was getting wrong info from the updated TS build from the monorepo.

For future reference: to avoid this, don't be a lazy ass like me and go run `yarn build` from the root folder.

:::tip

Does this mean you don't have to write an RPC server like in [Step #2](#step-2---write-an-rpc-server)?

> You don't technically have to, but then how would you run `nargo test` or `nargo prove`? To use both `Nargo` and `NoirJS` in your development flow, you will most certainly have to write a JSON RPC server anyway.

Reviewer: less :slightly_smiling_face: 
```suggestion
You don't technically have to, but then how would you run `nargo test` or `nargo prove`? To use both `Nargo` and `NoirJS` in your development flow, you will have to write a JSON RPC server.
```

Author: Lessssss 😈

(we should put that as #2 rule of devrel - #1 will always be "tag a DevRel")

:::

> In this case, let's make `foreignCallHandler` call the JSON RPC Server we created in [Step #2](#step-2---write-an-rpc-server), by making it a JSON RPC Client. For example, using the same `getSquared` circuit in [Step #1](#step-1---modify-your-circuit) (comments in the code):

Reviewer 2: ```suggestion
In this case, let's make `foreignCallHandler` call the JSON RPC Server we created in [Step #2](#step-2---write-an-rpc-server), by making it a JSON RPC Client.

For example, using the same `getSquared` program in [Step #1](#step-1---modify-your-noir-program) (comments in the code):
```

---
title: How to use Oracles
description: Learn how to use oracles in your Noir program with examples in both Nargo and NoirJS. This guide also covers writing a JSON RPC server and providing custom foreign call handlers for NoirJS.
keywords:
  - Noir Programming
  - Oracles
  - Nargo
  - NoirJS
  - JSON RPC Server
  - Foreign Call Handlers
sidebar_position: 1
---

This guide shows you how to use oracles in your Noir program. For the sake of clarity, it assumes that:

- You have read the [explainer on Oracles](../explainers/explainer-oracle.md) and are comfortable with the concept.
- You have a Noir program to add oracles to. You can create one using the [vite-hardhat starter](https://github.com/noir-lang/noir-starter/tree/main/vite-hardhat) as a boilerplate.
- You understand the concept of a JSON-RPC server. Visit the [JSON-RPC website](https://www.jsonrpc.org/) if you need a refresher.
- You are comfortable with server-side JavaScript (e.g. Node.js, managing packages, etc.).

> For reference, you can find the snippets used in this tutorial on the [Aztec DevRel Repository](https://github.com/AztecProtocol/dev-rel/tree/main/how_to_oracles/code-snippets/how-to-oracles).

Reviewer 2: Link to code is dead and could use an update.

Author: This is intended, as it points to an unmerged PR in the devrel repo

Reviewer 2: Thanks Ze. For the record, the PR is https://github.com/AztecProtocol/dev-rel/pull/118.

An example of an oracle that returns a `Field` would be:

```rust
#[oracle(getSqrt)]
unconstrained fn sqrt(number: Field) -> Field { }

unconstrained fn get_sqrt(number: Field) -> Field {
    sqrt(number)
}
```

In this example, we're wrapping our oracle function in a unconstrained method, and decorating it with `oracle(getSqrt)`. We can then call the unconstrained function as we would call any other function:

```rust
fn main(input: Field) {
    let sqrt = get_sqrt(input);
}
```

In the next section, we will make this `getSqrt` (defined on the `sqrt` decorator) be a method of the RPC server Noir will use.

:::danger

As explained in the [Oracle Explainer](../explainers/explainer-oracle.md), this `main` function is unsafe unless you constrain its return value. For example:

```rust
fn main(input: Field) {
    let sqrt = get_sqrt(input);

>     assert(sqrt[0].pow_32(2) as u64 == input as u64); // <---- constrain the return of an oracle!

Reviewer 3: I think this has to be just sqrt.pow_32!
